csrc/preseg_passes/remove_bcast_squeeze.cpp

```diff
@@ -8,45 +8,242 @@
+    Val* replacement = nullptr;
+    if (simplified_ops == first_ops) {
+      // The second op was simply a "Set" operation, so we just skip it
+      replacement = first->output(0);
+    } else {
+      switch (simple_op_type_opt.value()) {
+        case AxisOp::PRESERVE:
+          // This is equivalent to a set Op
+          replacement = input_tv;
+          break;
+        case AxisOp::SQUEEZE:
+          replacement = squeeze(input_tv, nonPreservedDims(simplified_ops));
+          break;
+        case AxisOp::BROADCAST:
+          replacement = broadcast(input_tv, nonPreservedDims(simplified_ops));
+          break;
+      }
+    }
+    NVF_ERROR(replacement != nullptr);
+    ir_utils::replaceValInAllExprInputsAndFusionOutputs(orig, replacement);
+    if (Expr* new_second = replacement->definition(); replacement != input_tv &&
+        isReplaceableExpr(new_second) &&
+        isReplaceableExpr(new_second->input(0)->definition())) {
+      // If we actually did a replacement, then we should process the
+      // replacement. Otherwise we would miss it.
+      maybeDoReplacement(new_second->input(0)->definition(), new_second);
+    }
+  }
+}
+
 // Remove broadcast-squeeze and squeeze-broadcast patterns
-// TODO: still remove when have intermediate ops between broadcast and squeeze
 void removeBcastSqueeze(Fusion* fusion) {
   // Iterate backwards over fusion expressions.
   // This will ensure we don't process expressions that are no longer valid
   // after replacement.
   auto exprs = fusion->exprs();
   for (auto it = std::rbegin(exprs); it != std::rend(exprs); it++) {
     Expr* expr = *it;
-    // step-1: find and remove broadcast + squeeze pattern
-    // before: Y = broadcast(X); Z = squeeze(Y);  M = someOp(Z)
-    // after : M = someOp(X)
-    // conditions: (1) broadcast and squeeze have the same dim flags
-    if (auto squeeze = dynamic_cast<SqueezeOp*>(expr)) {
-      if (auto bcast =
-              dynamic_cast<BroadcastOp*>(squeeze->in()->definition())) {
-        if (bcast->getBroadcastDimFlags() == squeeze->getSqueezeDimFlags()) {
-          ir_utils::replaceValInAllExprInputsAndFusionOutputs(
-              squeeze->out(), bcast->in());
-        }
-      }
-    }
-
-    // step-2: find and remove squeeze + broadcast pattern
-    // before: Y = squeeze(X); Z = broadcast(Y);  M = someOp(Z)
-    // after : M = someOp(X)
-    // conditions: (1) broadcast and squeeze have the same dim flags
-    if (auto bcast = dynamic_cast<BroadcastOp*>(expr)) {
-      if (auto squeeze = dynamic_cast<SqueezeOp*>(bcast->in()->definition())) {
-        if (bcast->getBroadcastDimFlags() == squeeze->getSqueezeDimFlags()) {
-          ir_utils::replaceValInAllExprInputsAndFusionOutputs(
-              bcast->out(), squeeze->in());
-        }
+    if (isReplaceableExpr(expr)) {
+      Expr* prev_expr = expr->input(0)->definition();
+      if (isReplaceableExpr(prev_expr)) {
+        maybeDoReplacement(prev_expr, expr);
```

Comment:
I don't like this recursion, but I wanted to check that this fixes our issue. It is needed, but I'd rather just stop using `fusion->exprs()` at the level up and instead manage our own stack-based DFS from the outputs at that level.